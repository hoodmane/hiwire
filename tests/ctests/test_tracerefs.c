#include "testlib.h"

#if defined(__wasi__)
int
extern_printf(const char*, ...)
  __attribute__((import_module("env"), import_name("printf")));

#define printf extern_printf
#endif

int
main()
{
  ASSERT(_hiwire_num_keys() == 0);
  HwRef id1 = hiwire_new(int_to_ref(5));
  ASSERT(_hiwire_num_keys() == 1);

  HwRef id2 = hiwire_new(int_to_ref(7));
  ASSERT(_hiwire_num_keys() == 2);
  HwRef id3 = hiwire_new(int_to_ref(10));
  ASSERT(_hiwire_num_keys() == 3);
  HwRef id4 = hiwire_new(int_to_ref(12));
  ASSERT(_hiwire_num_keys() == 4);

  ASSERT(ref_to_int(hiwire_get(id1)) == 5);
  ASSERT(ref_to_int(hiwire_get(id2)) == 7);
  ASSERT(ref_to_int(hiwire_get(id3)) == 10);
  ASSERT(ref_to_int(hiwire_get(id4)) == 12);

  hiwire_decref(id1);
  ASSERT(_hiwire_num_keys() == 3);
  ASSERT(is_null(hiwire_get(id1)));
  ASSERT(ref_to_int(hiwire_get(id2)) == 7);
  ASSERT(ref_to_int(hiwire_get(id3)) == 10);
  ASSERT(ref_to_int(hiwire_get(id4)) == 12);
  hiwire_decref(id3);
  ASSERT(is_null(hiwire_get(id1)));
  ASSERT(ref_to_int(hiwire_get(id2)) == 7);
  ASSERT(is_null(hiwire_get(id3)));
  ASSERT(ref_to_int(hiwire_get(id4)) == 12);
  ASSERT(_hiwire_num_keys() == 2);
  hiwire_decref(id2);
  ASSERT(_hiwire_num_keys() == 1);
  hiwire_decref(id4);
  ASSERT(_hiwire_num_keys() == 0);
  ASSERT(is_null(hiwire_get(id1)));
  ASSERT(is_null(hiwire_get(id2)));
  ASSERT(is_null(hiwire_get(id3)));
  ASSERT(is_null(hiwire_get(id4)));

  return 0;
}
