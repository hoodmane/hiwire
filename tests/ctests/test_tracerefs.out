[94;1mPASS [0m   _hiwire_num_keys() == 0
hiwire traceref { type: 'new', ref: 3, index: 1, value: 5, refcount: 1 }
[94;1mPASS [0m   _hiwire_num_keys() == 1
hiwire traceref { type: 'new', ref: 5, index: 2, value: 7, refcount: 1 }
[94;1mPASS [0m   _hiwire_num_keys() == 2
hiwire traceref { type: 'new', ref: 7, index: 3, value: 10, refcount: 1 }
[94;1mPASS [0m   _hiwire_num_keys() == 3
hiwire traceref { type: 'new', ref: 9, index: 4, value: 12, refcount: 1 }
[94;1mPASS [0m   _hiwire_num_keys() == 4
[94;1mPASS [0m   ref_to_int(hiwire_get(id1)) == 5
[94;1mPASS [0m   ref_to_int(hiwire_get(id2)) == 7
[94;1mPASS [0m   ref_to_int(hiwire_get(id3)) == 10
[94;1mPASS [0m   ref_to_int(hiwire_get(id4)) == 12
hiwire traceref { type: 'decref', ref: 3, index: 1, value: 5, refcount: 1 }
[94;1mPASS [0m   _hiwire_num_keys() == 3
[94;1mPASS [0m   is_null(hiwire_get(id1))
[94;1mPASS [0m   ref_to_int(hiwire_get(id2)) == 7
[94;1mPASS [0m   ref_to_int(hiwire_get(id3)) == 10
[94;1mPASS [0m   ref_to_int(hiwire_get(id4)) == 12
hiwire traceref { type: 'decref', ref: 7, index: 3, value: 10, refcount: 1 }
[94;1mPASS [0m   is_null(hiwire_get(id1))
[94;1mPASS [0m   ref_to_int(hiwire_get(id2)) == 7
[94;1mPASS [0m   is_null(hiwire_get(id3))
[94;1mPASS [0m   ref_to_int(hiwire_get(id4)) == 12
[94;1mPASS [0m   _hiwire_num_keys() == 2
hiwire traceref { type: 'decref', ref: 5, index: 2, value: 7, refcount: 1 }
[94;1mPASS [0m   _hiwire_num_keys() == 1
hiwire traceref { type: 'decref', ref: 9, index: 4, value: 12, refcount: 1 }
[94;1mPASS [0m   _hiwire_num_keys() == 0
[94;1mPASS [0m   is_null(hiwire_get(id1))
[94;1mPASS [0m   is_null(hiwire_get(id2))
[94;1mPASS [0m   is_null(hiwire_get(id3))
[94;1mPASS [0m   is_null(hiwire_get(id4))
